Kernel/Arch/x86/common/Spinlock.cpp

```diff
@@ -8,35 +8,31 @@
 
 namespace Kernel {
 
-u32 Spinlock::lock()
+InterruptsState Spinlock::lock()
 {
-    u32 prev_flags = cpu_flags();
+    InterruptsState interrupts_state_before_locking = Processor::interrupts_state();
     Processor::enter_critical();
-    cli();
+    Processor::disable_interrupts();
     while (m_lock.exchange(1, AK::memory_order_acquire) != 0)
         Processor::wait_check();
     track_lock_acquire(m_rank);
-    return prev_flags;
+    return interrupts_state_before_locking;
 }
 
-void Spinlock::unlock(u32 prev_flags)
+void Spinlock::unlock(InterruptsState interrupts_state_to_restore)
 {
     VERIFY(is_locked());
     track_lock_release(m_rank);
     m_lock.store(0, AK::memory_order_release);
 
     Processor::leave_critical();
-
-    if ((prev_flags & 0x200) != 0)
-        sti();
-    else
-        cli();
+    Processor::restore_interrupts_state(interrupts_state_to_restore);
 }
 
-u32 RecursiveSpinlock::lock()
+InterruptsState RecursiveSpinlock::lock()
 {
-    u32 prev_flags = cpu_flags();
-    cli();
+    InterruptsState interrupts_state_before_locking = Processor::interrupts_state();
+    Processor::disable_interrupts();
```

Comment:
Wouldn't it make sense to instead have `Processor::disable_interrupts()` return the previous state directly?

Reply:
Hmm, yeah it makes some sense, but it might not be immediately clear from the function name (`disable_interrupts`) that it also returns the previous interrupt state. I think for now we can keep them as two separate functions.